### PR TITLE
Generators startup hypothesis

### PR DESCRIPTION
**Pull request by ChamseddineBhd**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*

**What is the current behavior?** *(You can also link to an open issue here)*

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: src/main/java/com/powsybl/network/hypothesis/GeneratorsStartupAlgorithm.java [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    }
+
+    // calculate zone production (imposed and available)
+    // set if a group is usable or not + set startedPower
+    // this method sort zone groups
+    double evaluateProd(StartupZone startupZone) {
+        final double[] pMaxAvailable = {0};
+        for (StartupGroup startupGroup : startupZone.getStartupGroups()) {
+            GeneratorStartup generatorStartupExtension = startupGroup.getGenerator().getExtension(GeneratorStartup.class);
+            if (generatorStartupExtension.getPredefinedActivePowerSetpoint() != Double.MAX_VALUE) {
```
Comment (jonenst) [inline src/main/java/com/powsybl/network/hypothesis/GeneratorsStartupAlgorithm.java]: can this be null ?

File: src/main/java/com/powsybl/network/hypothesis/GeneratorsStartupAlgorithm.java [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+            }
+        }
+        return pMaxAvailable[0];
+    }
+
+    double evaluateAvailableMaxPower(StartupGroup startupGroup) {
+        double pMaxAvailable = 0;
+        // check if the group should started at pMax in parameters
+        GeneratorStartup ext = startupGroup.getGenerator().getExtension(GeneratorStartup.class);
+        double plannedOutageRate = ext != null ? ext.getPlannedOutageRate() : -1;
```
Comment (jonenst) [inline src/main/java/com/powsybl/network/hypothesis/GeneratorsStartupAlgorithm.java]: null is checked here but not above

File: src/main/java/com/powsybl/network/hypothesis/GeneratorsStartupAlgorithm.java [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    double hydroBandSetting = 0.1;
+
+    double lossCoefficient = 0;
+    double defaultAbatementCoefficient = 0;
+
+    StartupMarginalGroupType startUpMarginalGroupType = StartupMarginalGroupType.CLASSIC;
+
+    //thresholds for voltage adjustment
+    double pThreshold = 0.0;
+    double qThreshold = 0.0;
```
Comment (miovd) [inline src/main/java/com/powsybl/network/hypothesis/GeneratorsStartupAlgorithm.java]: Are those configurable?

File: src/main/java/com/powsybl/network/hypothesis/StartupGroup.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+/**
+ * @author Chamseddine BENHAMED <chamseddine.benhamed at rte-france.com>
+ */
+@Getter
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
```
Comment (miovd) [inline src/main/java/com/powsybl/network/hypothesis/StartupGroup.java]: Just a remark, but lombok is used nowhere else in powsybl, don't know if we should use it here either?

File: src/main/java/com/powsybl/network/hypothesis/StartupMarginalGroupType.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+ * Copyright (c) 2021, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.powsybl.network.hypothesis;
+
+/**
+ * @author Chamseddine BENHAMED <chamseddine.benhamed at rte-france.com>
+ */
```
Comment (miovd) [inline src/main/java/com/powsybl/network/hypothesis/StartupMarginalGroupType.java]: Could you document here in the javadoc what `MEXICO` and  `CLASSIC` refer to?

File: src/main/java/com/powsybl/network/hypothesis/StartupType.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+ * Copyright (c) 2021, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.powsybl.network.hypothesis;
+
+/**
+ * @author Chamseddine BENHAMED <chamseddine.benhamed at rte-france.com>
+ */
```
Comment (miovd) [inline src/main/java/com/powsybl/network/hypothesis/StartupType.java:11]: Could you document here in the javadoc what `EMPIL_ECO` and `POURCENT_PIMP` refer to?

File: src/main/java/com/powsybl/network/hypothesis/StartupZone.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import java.util.List;
+
+/**
+ * @author Chamseddine BENHAMED <chamseddine.benhamed at rte-france.com>
+ */
+@Getter
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
+@Builder
```
Comment (miovd) [inline src/main/java/com/powsybl/network/hypothesis/StartupZone.java]: Same remark about lombok

File: src/main/java/com/powsybl/network/hypothesis/GeneratorsStartupAlgorithm.java [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+
+    public void apply(Network network, StartupMarginalGroupType startUpMarginalGroupType, double defaultAbatementCoefficient,
```
Comment (miovd) [inline src/main/java/com/powsybl/network/hypothesis/GeneratorsStartupAlgorithm.java]: Maybe add a javadoc to explain what this algorithm does

File: src/main/java/com/powsybl/network/hypothesis/GeneratorsStartupAlgorithm.java (hunk 3 above)
Comment (ChamseddineBhd) [reply src/main/java/com/powsybl/network/hypothesis/GeneratorsStartupAlgorithm.java]: yes they are, we pass configuration values through the constructor

File: pom.xml [hunk 9]
```diff
@@ -66,11 +72,17 @@
...
 
         <powsyblcore.version>4.0.1</powsyblcore.version>
     </properties>
 
     <dependencies>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <version>${lombok.version}</version>
+        </dependency>
```
Comment (miovd) [inline pom.xml]: Can be deleted (not used)

File: src/main/java/com/powsybl/network/hypothesis/GeneratorsStartupAlgorithm.java [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+ * @author Chamseddine BENHAMED <chamseddine.benhamed at rte-france.com>
+ */
+public class GeneratorsStartupAlgorithm {
+    private static final Logger LOGGER = LoggerFactory.getLogger(GeneratorsStartupAlgorithm.class);
+
+    List<Generator> startupGroupsPowerMax = new ArrayList<>();
+
+    double nuclearBandSetting = 0.05;
+    double thermalBandSetting = 0.05;
+    double hydroBandSetting = 0.1;
```
Comment (miovd) [inline src/main/java/com/powsybl/network/hypothesis/GeneratorsStartupAlgorithm.java]: Maybe some comment here to indicate what these default values correspond to. Should they be configurable? (if yes, that can be done later)

File: src/main/java/com/powsybl/network/hypothesis/GeneratorsStartupAlgorithm.java [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+    double defaultAbatementCoefficient = 0;
+
+    StartupMarginalGroupType startUpMarginalGroupType = StartupMarginalGroupType.CLASSIC;
+
+    //thresholds for voltage adjustment
+    double pThreshold = 0.0;
+    double qThreshold = 0.0;
+
+    /**
+     * starts generators groups using Classic or Mexico algorithm
```
Comment (miovd) [inline src/main/java/com/powsybl/network/hypothesis/GeneratorsStartupAlgorithm.java]: It would be great to define in what context that can be useful and the unfolding of the algorithm in plain English

File: src/main/java/com/powsybl/network/hypothesis/StartupGroup.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.powsybl.network.hypothesis;
+
+import com.powsybl.iidm.network.Generator;
+
+/**
+ * @author Chamseddine BENHAMED <chamseddine.benhamed at rte-france.com>
```
Comment (miovd) [inline src/main/java/com/powsybl/network/hypothesis/StartupGroup.java]: If would be great to add a description of what this class models

File: src/main/java/com/powsybl/network/hypothesis/StartupGroupComparator.java [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.powsybl.network.hypothesis;
+
+import java.util.Comparator;
+
+/**
+ * @author Chamseddine BENHAMED <chamseddine.benhamed at rte-france.com>
```
Comment (miovd) [inline src/main/java/com/powsybl/network/hypothesis/StartupGroupComparator.java]: Quickly state in the javadoc how this comparator works

File: src/main/java/com/powsybl/network/hypothesis/StartupZone.java [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+ */
+package com.powsybl.network.hypothesis;
+
+import com.powsybl.iidm.network.Country;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * @author Chamseddine BENHAMED <chamseddine.benhamed at rte-france.com>
```
Comment (miovd) [inline src/main/java/com/powsybl/network/hypothesis/StartupZone.java]: Quickly state what this class models/what it is for

File: src/main/java/com/powsybl/network/hypothesis/StartupZoneBuilder.java [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+ */
+package com.powsybl.network.hypothesis;
+
+import com.powsybl.iidm.network.Country;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * @author Chamseddine BENHAMED <chamseddine.benhamed at rte-france.com>
```
Comment (miovd) [inline src/main/java/com/powsybl/network/hypothesis/StartupZoneBuilder.java]: Same comment

File: src/main/java/com/powsybl/network/hypothesis/GeneratorsStartupAlgorithm.java [hunk 16]
```diff
@@ -0,0 +1,274 @@
...
+                      double lossCoefficient, double pThreshold, double qThreshold, List<Generator> startupGroupsPowerMax) {
+        //parameters initialization
+        this.lossCoefficient = lossCoefficient;
+        this.defaultAbatementCoefficient = defaultAbatementCoefficient;
+        this.pThreshold = pThreshold;
+        this.qThreshold = qThreshold;
+        this.startUpMarginalGroupType = startUpMarginalGroupType;
+        this.startupGroupsPowerMax = startupGroupsPowerMax;
+
+        // prepare startupGroupsPerConnectedComponent HashMap
```
Comment (miovd) [inline src/main/java/com/powsybl/network/hypothesis/GeneratorsStartupAlgorithm.java]: It is more of an initialization to 0, isn't it?

File: src/main/java/com/powsybl/network/hypothesis/GeneratorsStartupAlgorithm.java [hunk 17]
```diff
@@ -0,0 +1,274 @@
...
+        Map<Component, List<StartupGroup>> startupGroupsPerConnectedComponent = prepareStartupGroupsPerConnectedComponentMap(network);
+
+        // for each connected component
+        startupGroupsPerConnectedComponent.keySet().forEach(component -> {
```
Comment (miovd) [inline src/main/java/com/powsybl/network/hypothesis/GeneratorsStartupAlgorithm.java]: Since you use key and value, you don't have to use `keySet()`, you can directly use `forEach`

File: src/main/java/com/powsybl/network/hypothesis/GeneratorsStartupAlgorithm.java [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+        return startupGroupsPerConnectedComponent;
+    }
+
+    // calculate  zone global consumption + losses
+    private void evaluateConsumption(Network network, StartupZone startupZone) {
+        final double[] consZone = {0};
+        final double[] consFictitious = {0};
+
+        network.getLoadStream().forEach(load -> {
+            if (load.getTerminal().getBusView().getBus().getConnectedComponent().getNum() == startupZone.getNum()) {
```
Comment (miovd) [inline src/main/java/com/powsybl/network/hypothesis/GeneratorsStartupAlgorithm.java]: Maybe check that `getBus` does not return `null` first or you will have a `NullPointerException`

**Review by miovd — APPROVED** [review]
